### DISCUSSION TITLE
dofs not found with get_dofs

### DISCUSSION THREAD

**Discussion by emi0000** [kinnala/scikit-fem#1144] [General]
Hi,

I just started with scikit fem and I'm not very clear about dofs. Taking a circle as mesh and selecting the boundary works as expected when using ElementTriP1() elements. However for ElementTriP2() elements not all dofs are found with get_dofs(). Are there different types of dofs? Thanks for comments. 

      import matplotlib.pyplot as plt
      import skfem as fem
      import skfem.visuals.matplotlib
      import numpy as np

      mesh = fem.MeshTri.init_circle(1)
      Vh = fem.Basis(mesh, fem.ElementTriP2())
      xy=Vh.doflocs

      plt.plot(xy[0],xy[1],'or')
      t=np.linspace(0,2*np.pi,1000)
      plt.plot(0.8*np.cos(t),0.8*np.sin(t),'--k')
      fem.visuals.matplotlib.draw(mesh, ax=plt.gca())

      def select(x): 
        gg=x[0]**2 + x[1]**2 > 0.8**2 
        return gg
 
      dofs = Vh.get_dofs(select)
      dofs.nodal

**Comment by emi0000** [comment]
sorry, I got it. The missing dofs are found in dofs.facet

**Comment by emi0000** [comment]
However, if I decrease the selection radius nodal nodes start to appear even though they are not in the selection region.  

```
mesh = fem.MeshTri.init_circle(1)
Vh = fem.Basis(mesh, fem.ElementTriP2())
xy=Vh.doflocs
def select(x): 
  gg=x[0]**2 + x[1]**2 > 0.72**2 
  return gg
 
dofs = Vh.get_dofs(select)
print(dofs.nodal,dofs.facet)

plt.plot(xy[0],xy[1],'or')
plt.plot(xy[0][dofs.nodal['u']],xy[1][dofs.nodal['u']],'Xk')
plt.plot(xy[0][dofs.facet['u']],xy[1][dofs.facet['u']],'Xg')
t=np.linspace(0,2*np.pi,1000)
plt.plot(0.72*np.cos(t),0.72*np.sin(t),'--k')
fem.visuals.matplotlib.draw(mesh, ax=plt.gca())
```

**Reply by kinnala** [reply]
The first (keyword) argument of get_dofs is called ’facets’. This will match facet midpoints. All DOFs on the matched facets including endpoint node DOFs will be returned.

**Reply by emi0000** [reply]
thanks,
so that means there will always be a difference between facets dofs and nodal dofs?
Is there a way to get all the dofs on the outer circumfence in one call in one list?

**Reply by kinnala** [reply]
Usually empty call works: Vh.get_dofs().all()
